src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -603,8 +603,30 @@ def _report_single_requirement_conflict(
 
         cands = self._finder.find_all_candidates(req.project_name)
         skipped_by_requires_python = self._finder.requires_python_skipped_reasons()
-        versions = [str(v) for v in sorted({c.version for c in cands})]
+        versions = [
+            str(v)
+            for v in sorted(
+                {
+                    c.version
+                    for c in cands
+                    if not (c.link.is_yanked if c.link else False)
+                }
+            )
+        ]
+        yanked_versions = [
+            str(v)
+            for v in sorted(
+                {c.version for c in cands if (c.link.is_yanked if c.link else False)}
+            )
+        ]
```

Comment:
Maybe we can compare `len(versions)` and `len(cands)` and skip building this list if they match (indicating there’re no yanked versions at all)

Reply:
Also I kind of wonder if the yanked version can be `info` instead. That’s just minor and subjective though.

Reply:
if I understood @pfmoore correctly, a version can be in both lists simultaneously if only some of the links for this version are yanked.

nothing to worry about for PyPI, but maybe this can be encountered elsewhere in the wild?

Reply:
An equivalent single-lookup variation:
```py
is_yanked = {True: set(), False: set()}
for c in cands:
    is_yanked[c.link.is_yanked if c.link else False].add(c.version)

versions = sorted(is_yanked[False])
yanked_versions = sorted(is_yanked[True])
```

How about that?

Reply:
Hi @uranusjr @pfmoore :wave:

I pushed the snippet above. Anything else from my side?